_integration-tests/tests/login_test.go

```diff
@@ -0,0 +1,166 @@
+}
+
+func (s *loginSuite) TestInvalidCredentialsError(c *check.C) {
+	err := s.writeCredentials(invalidLoginName)
+	c.Assert(err, check.IsNil, check.Commentf("error writting credentials"))
+
+	expectedMsg := "invalid credentials"
+	err = wait.ForFunction(c, expectedMsg, func() (string, error) { return s.stdout.String(), err })
+	c.Assert(err, check.IsNil, check.Commentf("didn't get expected invalid credentials error"))
+
+}
+
+func (s *loginSuite) TestFakeServerIsDetected(c *check.C) {
+	s.setUpHTTPServer()
+	s.setUpIPTables(c)
+	defer s.tearDownIPTables(c)
+	defer s.tearDownHTTPServer()
+
+	err := s.writeCredentials(validLoginName)
+	c.Assert(err, check.IsNil, check.Commentf("error writting credentials"))
+
+	expectedMsg := fmt.Sprintf("Post https://%s/api/v2/tokens/oauth: x509: certificate is valid for example.com, not %s", loginHost, loginHost)
+	err = wait.ForFunction(c, expectedMsg, func() (string, error) { return s.stdout.String(), err })
+	c.Assert(err, check.IsNil, check.Commentf("didn't get expected fake server error"))
+}
+
+func (s *loginSuite) handler(w http.ResponseWriter, r *http.Request) {
+	log.Panicf("in fake server handler, we should never get this far")
+}
+
+func (s *loginSuite) setUpHTTPServer() {
+	s.server = httptest.NewTLSServer(http.HandlerFunc(s.handler))
+
+	// URL is of the form https://ipaddr:port without trailing slash
+	s.serverAddrPort = strings.TrimLeft(s.server.URL, "https://")
+}
+
+func (s *loginSuite) setUpIPTables(c *check.C) {
+	cmd := ipTablesAddCommand(s.serverAddrPort)
+	_, err := cli.ExecCommandErr(cmd...)
+	c.Assert(err, check.IsNil, check.Commentf("Error setting up iptables"))
+}
+
+func (s *loginSuite) tearDownHTTPServer() {
+	s.server.Close()
+}
+
+func (s *loginSuite) tearDownIPTables(c *check.C) {
+	cmd := ipTablesDelCommand(s.serverAddrPort)
+	_, err := cli.ExecCommandErr(cmd...)
+	c.Assert(err, check.IsNil, check.Commentf("Error tearing down iptables"))
+}
+
+func ipTablesAddCommand(serverAddrPort string) []string {
+	return ipTablesCommand("A", serverAddrPort)
+}
+
+func ipTablesDelCommand(serverAddrPort string) []string {
+	return ipTablesCommand("D", serverAddrPort)
+}
+
+// action can be A for adding or D for deleting
+func ipTablesCommand(action, serverAddrPort string) []string {
+	return []string{"sudo",
+		"iptables", "-t", "nat", "-" + action, "OUTPUT", "-p", "tcp",
+		"-d", loginHost, "--dport", "443", "-j", "DNAT",
+		"--to-destination", serverAddrPort}
+}
+
+func (s *loginSuite) writeCredentials(loginName string) (err error) {
+	cmd := exec.Command("snappy", "login", loginName)
+	f, err := pty.Start(cmd)
+	if err != nil {
+		return err
+	}
+
+	err = cmdInteract(f, "Password: ", password)
+	if err != nil {
+		return err
+	}
+	io.Copy(&s.stdout, f)
+	return
+}
+
+func cmdInteract(f *os.File, prompt, input string) (err error) {
+	buf := make([]byte, 80)
+
+	n, err := f.Read(buf)
+	if err != nil {
+		return err
+	}
+	buf = buf[:n]
```

Comment:
This logic doesn't seem entirely right. An *os.File can break down the reads arbitrarily, so cutting down a read purely on how much was buffered doesn't have to match any particular expectation of length, which is what the logic below enforces.

Please reach me out online so we can come up with something, depending on what your intention is.